### PR TITLE
Support caching request FindMissingBlobs digests in the proxy

### PR DESCRIPTION
**Pull request by iain-macdonald**
This is the simplest possible implementation of this cache: an `lru.LRU` that sits in the `ContentAddressableStorageServerProxy` in which digests that exist in the remote cache are stored for a flag-configurable period of time. Because this approach is simple, it has some downsides, specifically:
1. This cache has no effect on the app's distributed cache traffic to serve these RPCs and
2. Digests are cached locally per-pod in proxy deployments instead of per deployment, so the reduction in traffic to the apps is less effective the larger the proxy deployment is (because each pod has to populate its local FindMissingBlobs cache).

I tried to come up with a way to address both of those problems. The first one is not too hard, we could do something similar within the distributed cache. The second is a little bit harder because we don't have a convenient mechanism with which we can populate the proxy's FindMissingBlob cache using responses from the remote. I thought we could put a FindMissingBlob cache in the distributed cache and then control it using a TTL in proxy deployments, but that means that FindMissingBlobs requests that miss that cache can't repopulate it, it would only be repopulated on reads from the backing cache.

I think that even with these caveats this is worth trying out to see how this cache performs. But I'll probably leave the bug open for a while so we can find this comment about the limitations in the future.

One last important thing to call out: this cache is sharded by group (among other properties) to handle partitions in the backing cache (which the proxy is unaware of) and encrypted groups.

Related issues: https://github.com/buildbuddy-io/buildbuddy-internal/issues/7780

### CODE REVIEW

**Comment by iain-macdonald** [comment]
> Adds an optional in-memory LRU in `CASServerProxy` caching digests reported present so repeat `FindMissingBlobs` requests within a TTL can be served locally. Clean, well-scoped, and safe-by-default (disabled unless TTL is set), with good comments and test coverage. Main items to address are the hash-only map keying edge case and a group-isolation test; the cache-consistency and atime trade-offs are inherent and already documented, so keep the rollout TTL conservative.
> 
> ### Additional findings (lines outside the diff)
> * `server/remote_cache/content_addressable_storage_server_proxy.go:203` — The previous code deliberately always proxied `FindMissingBlobs` "to avoid possible cache-inconsistency bugs"; this is now conditional. With caching on, if a blob is evicted from the remote within the TTL, the proxy will keep reporting it present → the client skips re-upload → a later read can fail. This is inherent to the feature and acknowledged, but worth confirming the intended TTL is short relative to remote eviction pressure and keeping this disabled until validated. No code change needed.
> * `server/remote_cache/content_addressable_storage_server_proxy.go:212` — The `cacheKeys` map is keyed by hash only, but keys encode size too. `DownloadString()` includes `size_bytes`, yet the map is keyed by `d.GetHash()`. If a single request contains two digests with the same hash but different sizes, the build loop's last write wins and both digests resolve to the same (wrong) cache key in the lookup/store loops (around lines 220 and 252). This is an edge case (valid CAS content-addressing makes it near-impossible), but keying the map by the digest proto (hash+size) or iterating without the intermediate map would make it robust. Low severity.
> * `server/remote_cache/content_addressable_storage_server_proxy.go:208` — A transient `AuthenticatedUser` error causes entries to be keyed under `AuthAnonymousUser`. This won't leak across groups (lookups for a real group use the real group ID and just miss), but it's worth a brief comment noting the fallback is intentional and non-leaking.
> * `server/remote_cache/content_addressable_storage_server_proxy.go:215` — On the hot path, a map plus a `DownloadString()` call is allocated per digest even when everything misses. Acceptable for the feature, and only incurred when caching is enabled.
> * `server/remote_cache/content_addressable_storage_server_proxy_test.go:316` — Not covered: per-group key sharding (two groups don't share cache entries) and eviction at `MaxSize`. A group-isolation test would be valuable given the encrypted-group motivation cited in the PR body.

I fixed 2 and 5. I think the others are either not fixable or not worth it right now.

File: enterprise/server/content_addressable_storage_server_proxy/content_addressable_storage_server_proxy.go [hunk 1]
```diff
@@ -162,9 +199,72 @@ func (s *CASServerProxy) FindMissingBlobs(ctx context.Context, req *repb.FindMis
...
-	// Always serve FindMissingBlobs requests out of the backing cache to
-	// avoid possible cache-inconsistency bugs.
-	return s.remote.FindMissingBlobs(ctx, req)
+	if s.findMissingCache == nil {
+		return s.remote.FindMissingBlobs(ctx, req)
+	}
+
+	// Construct all of the FindMissingBlobs cache keys upfront. We will need
+	// all of them, some of them twice.
+	groupID := interfaces.AuthAnonymousUser
```
Comment (tyler-french) [inline enterprise/server/content_addressable_storage_server_proxy/content_addressable_storage_server_proxy.go]: nit: looks like this line is not used

File: enterprise/server/content_addressable_storage_server_proxy/content_addressable_storage_server_proxy.go [hunk 2]
```diff
@@ -27,14 +28,33 @@ import (
 	repb "github.com/buildbuddy-io/buildbuddy/proto/remote_execution"
 )
 
-var enableGetTreeCaching = flag.Bool("cache_proxy.enable_get_tree_caching", false, "If true, the Cache Proxy attempts to serve GetTree requests out of the local cache. If false, GetTree requests are always proxied to the remote, authoritative cache.")
+var (
+	enableGetTreeCaching             = flag.Bool("cache_proxy.enable_get_tree_caching", false, "If true, the Cache Proxy attempts to serve GetTree requests out of the local cache. If false, GetTree requests are always proxied to the remote, authoritative cache.")
+	findMissingBlobsCacheTTL         = flag.Duration("cache_proxy.find_missing_blobs_cache_ttl", 0, "If greater than 0, the proxy caches digests reported by the backing cache as 'present' locally for FindMissingBlobs requests for this long.")
```
Comment (tyler-french) [inline enterprise/server/content_addressable_storage_server_proxy/content_addressable_storage_server_proxy.go:33]: totally optional: Do you think it would be worth having this a flagd flag? Or maybe we have both? This way we can adjust this based on traffic and disable quickly if there's an issue. I do think the cache size is good on startup. We also may want to add to the doc here that this needs to be << the app's CAS TTL.

File: enterprise/server/content_addressable_storage_server_proxy/content_addressable_storage_server_proxy.go [hunk 3]
```diff
@@ -162,9 +199,72 @@ func (s *CASServerProxy) FindMissingBlobs(ctx context.Context, req *repb.FindMis
...
+	if s.findMissingCache == nil {
+		return s.remote.FindMissingBlobs(ctx, req)
+	}
+
+	// Construct all of the FindMissingBlobs cache keys upfront. We will need
+	// all of them, some of them twice.
+	groupID := interfaces.AuthAnonymousUser
+	user, err := s.authenticator.AuthenticatedUser(ctx)
+	if err != nil {
+		log.Warningf("Error authenticating user, skipping FindMissingBlobs cache: %v", err)
```
Comment (tyler-french) [inline enterprise/server/content_addressable_storage_server_proxy/content_addressable_storage_server_proxy.go:213]: Will this warn on every call from an `ANON` user? If so this might cause a lot of noise. We may want to log only if `!authutil.IsAnonymousUserError(err)`

File: enterprise/server/content_addressable_storage_server_proxy/content_addressable_storage_server_proxy.go [hunk 4]
```diff
@@ -162,9 +199,72 @@ func (s *CASServerProxy) FindMissingBlobs(ctx context.Context, req *repb.FindMis
...
+	rsp, err := s.remote.FindMissingBlobs(ctx, remoteReq)
+	if err != nil {
+		return rsp, err
+	}
+
+	// Cache the blobs the backing cache reported as present. We need to invert
+	// the set because the remote reports what's missing, not what's present.
+	missing := make(map[string]struct{}, len(rsp.GetMissingBlobDigests()))
+	for _, d := range rsp.GetMissingBlobDigests() {
+		missing[d.GetHash()] = struct{}{}
```
Comment (tyler-french) [inline enterprise/server/content_addressable_storage_server_proxy/content_addressable_storage_server_proxy.go]: clarification: Should we use `digestKey` here?

File: enterprise/server/content_addressable_storage_server_proxy/content_addressable_storage_server_proxy.go [hunk 5]
```diff
@@ -162,9 +199,72 @@ func (s *CASServerProxy) FindMissingBlobs(ctx context.Context, req *repb.FindMis
...
+	// Consult the local FindMissingBlobs cache first to remove some digests.
+	misses := make([]*repb.Digest, 0, len(req.GetBlobDigests()))
+	for _, d := range req.GetBlobDigests() {
+		if !s.findMissingCache.Contains(cacheKeys[digestKey(d)]) {
+			misses = append(misses, d)
+		}
+	}
+
+	hits := len(req.GetBlobDigests()) - len(misses)
+	s.findMissingCacheHits.Add(float64(hits))
```
Comment (tyler-french) [inline enterprise/server/content_addressable_storage_server_proxy/content_addressable_storage_server_proxy.go:230]: Not worth adjusting the cache implementation here, but could we attach `cdc.IsChunked(ctx)` to the metrics to see how effective this is based on type of traffic chunks vs. blobs? This header is only set from the executor and the proxies for now (not bazel yet), but I think it will still be interesting to see.

File: enterprise/server/content_addressable_storage_server_proxy/content_addressable_storage_server_proxy.go (hunk 1 above)
Comment (iain-macdonald) [reply enterprise/server/content_addressable_storage_server_proxy/content_addressable_storage_server_proxy.go]: Whoops, done.

File: enterprise/server/content_addressable_storage_server_proxy/content_addressable_storage_server_proxy.go (hunk 4 above)
Comment (iain-macdonald) [reply enterprise/server/content_addressable_storage_server_proxy/content_addressable_storage_server_proxy.go]: Yeah, fixed.

File: enterprise/server/content_addressable_storage_server_proxy/content_addressable_storage_server_proxy.go (hunk 3 above)
Comment (iain-macdonald) [reply enterprise/server/content_addressable_storage_server_proxy/content_addressable_storage_server_proxy.go:213]: Updated to not log in that case, and store the FMB digest in the cache as well.

File: enterprise/server/content_addressable_storage_server_proxy/content_addressable_storage_server_proxy.go (hunk 2 above)
Comment (iain-macdonald) [reply enterprise/server/content_addressable_storage_server_proxy/content_addressable_storage_server_proxy.go:33]: I don't think we want to be adjusting the size of this cache while the process is running, so any updates will require a restart. Given that, it seems fine to just leave it as a command-line flag.

File: enterprise/server/content_addressable_storage_server_proxy/content_addressable_storage_server_proxy.go (hunk 5 above)
Comment (iain-macdonald) [reply enterprise/server/content_addressable_storage_server_proxy/content_addressable_storage_server_proxy.go:230]: If that's stored in the context, what happens if some of the digests in batch are chunks and some aren't?

**Review by tylerwilliams — APPROVED** [review]

File: enterprise/server/content_addressable_storage_server_proxy/content_addressable_storage_server_proxy.go (hunk 5 above)
Comment (tyler-french) [reply enterprise/server/content_addressable_storage_server_proxy/content_addressable_storage_server_proxy.go:230]: The header tells the server to skip checking individual blobs for if they're composed of chunks. 

We could somehow do a two layer cache here, where chunked traffic checks whole blobs cache, then non chunked checks both layers, or we could add this as a bit in the key (but this might double the size).

Its not a big deal for now, we'd just need to disable or fix this before changing chunk sizes, since that's the only time the issue surfaces.
